modules/admin/src/test/java/org/motechproject/admin/it/AdminBundleIT.java

```diff
@@ -114,6 +127,48 @@ public void testMessageController() throws IOException, InterruptedException {
         assertTrue("No messages listed", json.size() > 0);
     }
 
+    @Test
+    public void testUploadBundleFromRepository() throws IOException, InterruptedException {
+        boolean isUploadSuccessful = uploadBundle("Repository", "org.motechproject:cms-lite:LATEST", null, "on");
+
+        assertTrue(isUploadSuccessful);
+    }
+
+    @Test
+    public void testUploadBundleFromFile() throws IOException, InterruptedException {
+        File file = new File("src/test/resources/motech-upload-test-bundle.jar");
+
+        boolean isUploadSuccessful = uploadBundle("File", null, file, "on");
+
+        assertTrue(isUploadSuccessful);
+    }
+
+    private boolean uploadBundle(String moduleSource, String moduleId, File bundleFile, String startBundle) throws IOException, InterruptedException {
+        String uri = String.format("http://%s:%d/admin/api/bundles/upload/", HOST, PORT);
+
+        HttpPost httpPost = new HttpPost(uri);
+        MultipartEntityBuilder entity = MultipartEntityBuilder.create();
+        Charset chars = Charset.forName("UTF-8");
+        entity.setCharset(chars);
+        entity.addTextBody("moduleSource", moduleSource, ContentType.MULTIPART_FORM_DATA);
+        if(moduleSource.equals("Repository")) entity.addTextBody("moduleId", moduleId, ContentType.MULTIPART_FORM_DATA);
+        else if(moduleSource.equals("File")) {
+            if(bundleFile == null) return false;
+            entity.addBinaryBody("bundleFile", bundleFile);
+        }
+        else return false;
+        entity.addTextBody("startBundle", startBundle, ContentType.MULTIPART_FORM_DATA);
+
+        httpPost.setEntity(entity.build());
+
+        int bundlesCountBeforeUpload = bundleContext.getBundles().length;
+        HttpResponse response = getHttpClient().execute(httpPost);
+        EntityUtils.consume(response.getEntity());
+        int bundlesCountAfterUpload = bundleContext.getBundles().length;
+
+        return (bundlesCountAfterUpload == bundlesCountBeforeUpload + 1);
```

Comment:
I wonder if this check is enough to verify that the bundles have properly started. What does this count return? If the bundle fails to start, is it included in the count? Should we be checking for the status of the added bundle additionaly?